From patchwork Wed Feb  3 01:28:40 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Krish Sadhukhan <krish.sadhukhan@oracle.com>
X-Patchwork-Id: 12063195
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-18.7 required=3.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,
	INCLUDES_PATCH,MAILING_LIST_MULTI,MSGID_FROM_MTA_HEADER,SPF_HELO_NONE,
	SPF_PASS,USER_AGENT_GIT autolearn=ham autolearn_force=no version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id B1E4AC433DB
	for <kvm@archiver.kernel.org>; Wed,  3 Feb 2021 02:17:56 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 79DBB64F50
	for <kvm@archiver.kernel.org>; Wed,  3 Feb 2021 02:17:56 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S232027AbhBCCRx (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Tue, 2 Feb 2021 21:17:53 -0500
Received: from aserp2120.oracle.com ([141.146.126.78]:49754 "EHLO
        aserp2120.oracle.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S230087AbhBCCRu (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 2 Feb 2021 21:17:50 -0500
Received: from pps.filterd (aserp2120.oracle.com [127.0.0.1])
        by aserp2120.oracle.com (8.16.0.42/8.16.0.42) with SMTP id
 1132FliO008298;
        Wed, 3 Feb 2021 02:16:50 GMT
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=oracle.com;
 h=from : to : cc :
 subject : date : message-id : in-reply-to : references :
 content-transfer-encoding : content-type : mime-version;
 s=corp-2020-01-29; bh=cOTE7Y8ZqD5+Y50rLYXL8ENXmTwJEN8O4esUDbNCmqc=;
 b=0dnuyx8gQu9oPUkKEoLyiMRuCk/v7Pl7pEWUVu1Jj3LrTdWUxZNWq5m0ThOcPTaemIo7
 1CSDOaxPv6FiW170AfKHHJpvW/wMnmTKgGMIeee1gELYCvzQIY9I9Ro5Iby7Gx6devpa
 7NAWUnmQvo6lC4wZ+Bzbs5cAgJoRvutUZoA5gCCalTUT9nezuk2Ac/UUiaNqHjFII5I8
 rOVPL1nEQeqqH585MDJhuM+zLsSy2w+yqctKdygZdfZ5NXsnr+ztaQMEeRBAeiLIZw/q
 3Xw29xoz5V8bWzOaMhw/YVV94C0LdEzXrhBHQI4fn+VsaM0MSc5Rz05TaO/yMqZ+k4yG FA==
Received: from userp3030.oracle.com (userp3030.oracle.com [156.151.31.80])
        by aserp2120.oracle.com with ESMTP id 36cydkwq0v-1
        (version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Wed, 03 Feb 2021 02:16:50 +0000
Received: from pps.filterd (userp3030.oracle.com [127.0.0.1])
        by userp3030.oracle.com (8.16.0.42/8.16.0.42) with SMTP id
 1132Eanb110778;
        Wed, 3 Feb 2021 02:16:49 GMT
Received: from nam10-bn7-obe.outbound.protection.outlook.com
 (mail-bn7nam10lp2102.outbound.protection.outlook.com [104.47.70.102])
        by userp3030.oracle.com with ESMTP id 36dhcxqdsa-1
        (version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Wed, 03 Feb 2021 02:16:49 +0000
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=RD60Wu6p2y1tr7Ig5Pq7hAOVDqu9xdVJnFOyMoqW/z5qEl2FQiZgWxE80FTXRyc7PfuhevnMjlxo5jqOi55iYZ+mZJA++bU1JndGou0zP10TrWXqmnu9rt9Cy8fft4DudOgbGiv20QxcrZxk2oiiyUpc3Rg/x4VtTD7fSS45ADNPzELuls+UReLOlDhLMYxOgb+Z3EfTXRCNIYtwpj6R5MmBduQhSjkcF9Gn3HjwBu8QzwIKOQtjwB4aPWclbq4/wWrgRA3Ia5hoft6lBsFTUHuMuk4lDB3uSr6wjfH3l8nzkDBI5OTQ4CC+gUln8TARnE5vsvj9qVMH5jmjViC0LQ==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=cOTE7Y8ZqD5+Y50rLYXL8ENXmTwJEN8O4esUDbNCmqc=;
 b=CzOVDn2qLYzMM5ws1zt7LtOzixC7ZbOQVdINo6MI+Jblblbtghjah0LUEOT/4msaKPwlmfNNAdgZPf3y42VOoqSLYbctj0CQSeqG0iDHuJYd0XOmOOUrfxN2JSFo+Wtb8H3x5B3WWSsRTFziX3C+VJSUc1R+kc87Gt0cD4wBEDhxxycCd7u9V5lRnxJAxVBM0PLv9cevm0UZIZEFSSYE7ZxJATE3HFFzkxJ+fPTFvouRxmk3ANsNj/v+MOJ/XfpmGD4VzmC9r6aBiUuu+uH6wC2wKyLfB5ptXhSxAI18XsO09rPlXtN5mESSkwPHjbdvzPyqQbf+HmYwOwUBpxLBjw==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=oracle.com; dmarc=pass action=none header.from=oracle.com;
 dkim=pass header.d=oracle.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=oracle.onmicrosoft.com; s=selector2-oracle-onmicrosoft-com;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=cOTE7Y8ZqD5+Y50rLYXL8ENXmTwJEN8O4esUDbNCmqc=;
 b=Vove83Fe9nof2JZ6gacMN5/dNlPMx7W218oUmvUCvGMttfa3kcALrV1Tn5ygu180a6ORZnoc84clO9K3UDJ8wNlfaEVGErtJ74GGcwaOe3WISym8U2kWh99bSIMbTIFh8N9BC9rQOOoD03Q/cbBf3daOBFp2IRn55jwUFQ7e7FU=
Authentication-Results: vger.kernel.org; dkim=none (message not signed)
 header.d=none;vger.kernel.org; dmarc=none action=none header.from=oracle.com;
Received: from SN6PR10MB3021.namprd10.prod.outlook.com (2603:10b6:805:cc::19)
 by SA2PR10MB4793.namprd10.prod.outlook.com (2603:10b6:806:110::24) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.3805.17; Wed, 3 Feb
 2021 02:16:47 +0000
Received: from SN6PR10MB3021.namprd10.prod.outlook.com
 ([fe80::1871:3741:cc17:bcf7]) by SN6PR10MB3021.namprd10.prod.outlook.com
 ([fe80::1871:3741:cc17:bcf7%7]) with mapi id 15.20.3825.020; Wed, 3 Feb 2021
 02:16:47 +0000
From: Krish Sadhukhan <krish.sadhukhan@oracle.com>
To: kvm@vger.kernel.org
Cc: pbonzini@redhat.com, jmattson@google.com, seanjc@google.com
Subject: [PATCH 1/3] KVM: SVM: Replace hard-coded value with #define
Date: Tue,  2 Feb 2021 20:28:40 -0500
Message-Id: <20210203012842.101447-2-krish.sadhukhan@oracle.com>
X-Mailer: git-send-email 2.25.4
In-Reply-To: <20210203012842.101447-1-krish.sadhukhan@oracle.com>
References: <20210203012842.101447-1-krish.sadhukhan@oracle.com>
X-Originating-IP: [138.3.200.29]
X-ClientProxiedBy: SJ0PR03CA0096.namprd03.prod.outlook.com
 (2603:10b6:a03:333::11) To SN6PR10MB3021.namprd10.prod.outlook.com
 (2603:10b6:805:cc::19)
MIME-Version: 1.0
X-MS-Exchange-MessageSentRepresentingType: 1
Received: from ban25x6uut29.us.oracle.com (138.3.200.29) by
 SJ0PR03CA0096.namprd03.prod.outlook.com (2603:10b6:a03:333::11) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.3825.19 via Frontend
 Transport; Wed, 3 Feb 2021 02:16:45 +0000
X-MS-PublicTrafficType: Email
X-MS-Office365-Filtering-Correlation-Id: 3f5d3c58-c0c2-4856-e69e-08d8c7e9c182
X-MS-TrafficTypeDiagnostic: SA2PR10MB4793:
X-Microsoft-Antispam-PRVS: 
 <SA2PR10MB47937B05D025DAA1670E274E81B49@SA2PR10MB4793.namprd10.prod.outlook.com>
X-MS-Oob-TLC-OOBClassifiers: OLM:3044;
X-MS-Exchange-SenderADCheck: 1
X-Microsoft-Antispam: BCL:0;
X-Microsoft-Antispam-Message-Info: 
 MXbeph7s57xoFdBIbLZZ733u3GlhL8kSydX8IK7eBgJngM/CbXoqWYH588+iZk7Ez59aiW0DeUDTSOzEw083msjLkfpsKGVUX4pNLYwIimzDB2yTtSea4Xb+VmLYI6MjJuzZOiLPLdEXUuoOKIQ0WQKSwouPJF5IPMgq/uAMFWQkMJxkMSW/KyAaXxXuXLSMIh78V0jCy6q6oWTW6KyWYl3l3wyjD9x55ja7AE9yqmkBNzxvT63uwgd8liLlBe7LmWZJIaDpTIbCbWTyBOic6Qa29Pq1EW3yso2ULqfT0IwSgRzp4pTBWqGXAK6QLi+Ybdn36v46VQG6X2ztdU2AoZxV0ty/ng8xnXF5kMQxd5rMUVjJd4UpkB5ZPPSsRJVoOusNEf4Mp+YGYgNiKKc3LaWh9FgWrwpSCE6LI4IjBQSvtau6+ZXOr5bOcy5bru5u+BoyiFOnIzuyubavFudRxBlZa/gvapw5QjpRlOh0QW03cDug+OwOFnKDlEIasXMxSEWZaKjuIzrWq5MW8/lkzQ==
X-Forefront-Antispam-Report: 
 CIP:255.255.255.255;CTRY:;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:SN6PR10MB3021.namprd10.prod.outlook.com;PTR:;CAT:NONE;SFS:(366004)(39860400002)(376002)(346002)(396003)(136003)(316002)(6666004)(36756003)(8676002)(66946007)(16526019)(66476007)(6486002)(7696005)(83380400001)(4326008)(6916009)(2906002)(86362001)(4744005)(5660300002)(956004)(186003)(1076003)(44832011)(52116002)(8936002)(26005)(478600001)(2616005)(66556008);DIR:OUT;SFP:1101;
X-MS-Exchange-AntiSpam-MessageData: 
 m+tKoxS5gadJjLYOwMRb4i8bV4kz+TAnzAljPQXZ/5GOXZ8o0r3pVVCFG4FN7kvLbtKrhH8zGMBPP/cx9uTWjJchGdvB5qcf5ehrPJfoNst2JDyCCZMnd+rJKEAc5GadehYEUH2c8MzcN+jr8bEKzVM8+7jNeH9fzrHfd8NmqFHez7YXcoghtGIbHXv8CLIO3c3qcUbntzmhFmhKbsV9kTuWUAxBGti1+RhJvzNFRhEsX/3ujkLYYhJW6OaVOMYDLgZZWHwYgCuCJfTkPRU0/k4hHQGE+quKiHT8jajst/BEwUCdVBdXwwJjMYNWv5taX51dNb1n0pH170p3TbbEXu6Wc4nK9RHaCr2JYb5e6av4Ub7vJ/X6XGC+hD+ifbwIBVryqhomfxIOFeTtQO+uNM1jtFeDHCbRqas1mXXVZBlFT4s9Pk2ojSRsgJBfJUfVdSEZnP4MDvDuEZBCR/CVinwAjiBPfxIz7PwdblYfxQythjdPJsnUuviF7IJy7eL4KF2dYoxoywbDTs2PH+3vayOeKAe9oqibEWmSfwbGd8Hq/GPN1+ydRx1iXLBheB+x1N/Rdi+eY3tEdYcSr23QLh6trYGp8x1BZYJSwdS3eSQaZil7Mh/ByQEnh7wBCY9iECHxxm0OcE6g/jwcVAQO3cz3C7kJEqP21Lh2e2zlRshHKKpbTPezArwW4b3GIuCOhQtorTjaMXdNbiCg5wVHMcxSag+9/YfWs/wkOCxxWVysQeLSPKg6zwuMHd0yaZjkSBMEloAEbwOYXnUaOrwlstMA2MJ8ZpPW44l+YwPxiY6cVvPHwNchNghOI/F4tkv40J6VZU6FWVZgGptHhmLTy6L/NhWpkQ2g060c8OtoWIc7oEWVUewDk1lbGTQP9CC7sNkF68SaC5C/Ks1kGvbAErQzBnc5DSsPczhdjqBttnPMikBw2K7MAOTCOKCU5Z4Sdg0+ITgtOetF+DPg0OaGjMuQaWf0ggsI+O9Za5GApJRjzZMrA77lP6dyz5K0MlV+R4Zys/CXiNsExB3TviJhU4ag5DcO+832xpHfMAgUvYeWL+XTdXk4h+ucR9MkjUce
X-OriginatorOrg: oracle.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 3f5d3c58-c0c2-4856-e69e-08d8c7e9c182
X-MS-Exchange-CrossTenant-AuthSource: SN6PR10MB3021.namprd10.prod.outlook.com
X-MS-Exchange-CrossTenant-AuthAs: Internal
X-MS-Exchange-CrossTenant-OriginalArrivalTime: 03 Feb 2021 02:16:47.1237
 (UTC)
X-MS-Exchange-CrossTenant-FromEntityHeader: Hosted
X-MS-Exchange-CrossTenant-Id: 4e2c6054-71cb-48f1-bd6c-3a9705aca71b
X-MS-Exchange-CrossTenant-MailboxType: HOSTED
X-MS-Exchange-CrossTenant-UserPrincipalName: 
 rbf+U6OkSJ8nvckq35fG2oeoPm8XdQuk3rd6SJhRqY2XATCOIyvde0Bu49og3N5ZGV8/wAD5c7v8y9iysL9Qgf10eIioppGnAV2n0BYII0Q=
X-MS-Exchange-Transport-CrossTenantHeadersStamped: SA2PR10MB4793
X-Proofpoint-Virus-Version: vendor=nai engine=6200 definitions=9883
 signatures=668683
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 mlxscore=0
 mlxlogscore=849 phishscore=0
 spamscore=0 suspectscore=0 malwarescore=0 adultscore=0 bulkscore=0
 classifier=spam adjust=0 reason=mlx scancount=1 engine=8.12.0-2009150000
 definitions=main-2102030007
X-Proofpoint-Virus-Version: vendor=nai engine=6200 definitions=9883
 signatures=668683
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 bulkscore=0
 adultscore=0
 priorityscore=1501 impostorscore=0 malwarescore=0 clxscore=1015
 spamscore=0 lowpriorityscore=0 phishscore=0 mlxlogscore=906 mlxscore=0
 suspectscore=0 classifier=spam adjust=0 reason=mlx scancount=1
 engine=8.12.0-2009150000 definitions=main-2102030007
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

Replace the hard-coded value for bit# 1 in EFLAGS, with the available
#define.

Signed-off-by: Krish Sadhukhan <krish.sadhukhan@oracle.com>
---
 arch/x86/kvm/svm/svm.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/x86/kvm/svm/svm.c b/arch/x86/kvm/svm/svm.c
index f923e14e87df..5435f5cb756b 100644
--- a/arch/x86/kvm/svm/svm.c
+++ b/arch/x86/kvm/svm/svm.c
@@ -1202,7 +1202,7 @@ static void init_vmcb(struct vcpu_svm *svm)
 
 	svm_set_efer(&svm->vcpu, 0);
 	save->dr6 = 0xffff0ff0;
-	kvm_set_rflags(&svm->vcpu, 2);
+	kvm_set_rflags(&svm->vcpu, X86_EFLAGS_FIXED);
 	save->rip = 0x0000fff0;
 	svm->vcpu.arch.regs[VCPU_REGS_RIP] = save->rip;
 

From patchwork Wed Feb  3 01:28:41 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Krish Sadhukhan <krish.sadhukhan@oracle.com>
X-Patchwork-Id: 12063197
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-18.7 required=3.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,
	INCLUDES_PATCH,MAILING_LIST_MULTI,MSGID_FROM_MTA_HEADER,SPF_HELO_NONE,
	SPF_PASS,URIBL_BLOCKED,USER_AGENT_GIT autolearn=ham autolearn_force=no
	version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id A9598C433E0
	for <kvm@archiver.kernel.org>; Wed,  3 Feb 2021 02:19:53 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 55F1964F59
	for <kvm@archiver.kernel.org>; Wed,  3 Feb 2021 02:19:53 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S231292AbhBCCTj (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Tue, 2 Feb 2021 21:19:39 -0500
Received: from aserp2120.oracle.com ([141.146.126.78]:51138 "EHLO
        aserp2120.oracle.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S229879AbhBCCTf (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 2 Feb 2021 21:19:35 -0500
Received: from pps.filterd (aserp2120.oracle.com [127.0.0.1])
        by aserp2120.oracle.com (8.16.0.42/8.16.0.42) with SMTP id
 1132EuGG006449;
        Wed, 3 Feb 2021 02:18:50 GMT
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=oracle.com;
 h=from : to : cc :
 subject : date : message-id : in-reply-to : references :
 content-transfer-encoding : content-type : mime-version;
 s=corp-2020-01-29; bh=ModOCWIXpk1rwpr2dJxzdupuucZ9YJ+Tx7/IW3ncBjo=;
 b=s7/2jkFdofFuSOiVFG2txppP3spNvPrBmHuT+YJiZIgGzUaSl7ZBvB4SIqUV7f2S3XW8
 LLVF/ygjzZEebn/1L3VodMyjaKUJKGRUUbKIHdUL3Io/zDIpj+VLwjjp33Ct89IZpF0S
 wz7B20UMbxSBZLy7oe6lu0JN+8LvLjnJA4e1mQRyW6IA5RbI4RDCxvqY/h81lsk0nPcU
 CZbAQGld0rqFmbpdKqeHtC8p1WRVL25s9nA8kvDz1mjgceEUuc2csE6evW+kGNxv+7km
 1tZEjyhom+HTrXWrN3YK0C7yQeYUNvRxSMHgVpMTaa/tEDnlrD5xeIR9bS3G+e6k+7su WQ==
Received: from userp3030.oracle.com (userp3030.oracle.com [156.151.31.80])
        by aserp2120.oracle.com with ESMTP id 36cydkwq46-1
        (version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Wed, 03 Feb 2021 02:18:50 +0000
Received: from pps.filterd (userp3030.oracle.com [127.0.0.1])
        by userp3030.oracle.com (8.16.0.42/8.16.0.42) with SMTP id
 1132Eanc110778;
        Wed, 3 Feb 2021 02:16:49 GMT
Received: from nam10-bn7-obe.outbound.protection.outlook.com
 (mail-bn7nam10lp2102.outbound.protection.outlook.com [104.47.70.102])
        by userp3030.oracle.com with ESMTP id 36dhcxqdsa-2
        (version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Wed, 03 Feb 2021 02:16:49 +0000
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=kJRwOY1NXsy4yKFLJxgEtz5TarN9SNputqNcW+188RzS2oUAwX3RqNhHr8pnBzk05DfJQDpcMqScKgQXgqdOL2jVIlcJAv8vSaj1iBq7A7nUnNVUufmfpPr30T20O9za04OgDZAP50p+recNfM7MIAk2/uMrzoHXtZewlPTFUKFSCccBJqXy2UjKDg0drbJI5zD4hg160KZSvY4h7aY8PVhU4K3tO9HgQIMI0lT93qcwyz6VUBsd+5ive6yoCQwFbsGFarSiT80JhR3Qt/qJ+UJA3prspPvEZgOHSEtycSK774GXWG/Ngn21bPYWFv92rj6ndpCJKnUBqlbe/yw3ZQ==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=ModOCWIXpk1rwpr2dJxzdupuucZ9YJ+Tx7/IW3ncBjo=;
 b=Ml8CW5M/KZocFyKR9Kw0LclwfwWi3BCE3jvy8jcSb/ZjW6Uy0PU7OsRdSw8KZ9bxFKDI/mPyYlZcBppw/WLkuP2GciNn+T0ZpNETvywWSKZRDpJbQBVE2ky+WfBUM8zh5I5IcpC/ICbMF5s7TtluH+m1/gJdaa2CnX3IuHodpZoTILAN5DnOo8YeoNYmH1sj9L//Ne0sGdDuqcfGvv79fU4UUH5hi+F5bC8IshIeCl+VXRuXV/xofIPCvXoSZHPvwbRB54B76hEn0KhQQQtBT8xUkHI857E2as8eUy1GattozjVH0e1K0mUEzmUFmmtxa5fp7KddVYyh4s6wM7raLg==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=oracle.com; dmarc=pass action=none header.from=oracle.com;
 dkim=pass header.d=oracle.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=oracle.onmicrosoft.com; s=selector2-oracle-onmicrosoft-com;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=ModOCWIXpk1rwpr2dJxzdupuucZ9YJ+Tx7/IW3ncBjo=;
 b=GQ+TiyenJVEBe0+0Z5Eb/hQ8gptCU/FH6JBryH8CL6MFLmgxJzZWWjgEQONewdIMKRx+Q4HluVC4wbwkv/Qgv18vlQYFyI5t4vgRykk9iaNBt+r35cWGTuML2uER6F2V4ccLwEhe2ZGsDUE1nKxsY7Y7hlkoXhB3NJNBQSSm4P4=
Authentication-Results: vger.kernel.org; dkim=none (message not signed)
 header.d=none;vger.kernel.org; dmarc=none action=none header.from=oracle.com;
Received: from SN6PR10MB3021.namprd10.prod.outlook.com (2603:10b6:805:cc::19)
 by SA2PR10MB4793.namprd10.prod.outlook.com (2603:10b6:806:110::24) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.3805.17; Wed, 3 Feb
 2021 02:16:48 +0000
Received: from SN6PR10MB3021.namprd10.prod.outlook.com
 ([fe80::1871:3741:cc17:bcf7]) by SN6PR10MB3021.namprd10.prod.outlook.com
 ([fe80::1871:3741:cc17:bcf7%7]) with mapi id 15.20.3825.020; Wed, 3 Feb 2021
 02:16:48 +0000
From: Krish Sadhukhan <krish.sadhukhan@oracle.com>
To: kvm@vger.kernel.org
Cc: pbonzini@redhat.com, jmattson@google.com, seanjc@google.com
Subject: [PATCH 2/3] nVMX: Add helper functions to set/unset host RFLAGS.TF on
 the VMRUN instruction
Date: Tue,  2 Feb 2021 20:28:41 -0500
Message-Id: <20210203012842.101447-3-krish.sadhukhan@oracle.com>
X-Mailer: git-send-email 2.25.4
In-Reply-To: <20210203012842.101447-1-krish.sadhukhan@oracle.com>
References: <20210203012842.101447-1-krish.sadhukhan@oracle.com>
X-Originating-IP: [138.3.200.29]
X-ClientProxiedBy: SJ0PR03CA0096.namprd03.prod.outlook.com
 (2603:10b6:a03:333::11) To SN6PR10MB3021.namprd10.prod.outlook.com
 (2603:10b6:805:cc::19)
MIME-Version: 1.0
X-MS-Exchange-MessageSentRepresentingType: 1
Received: from ban25x6uut29.us.oracle.com (138.3.200.29) by
 SJ0PR03CA0096.namprd03.prod.outlook.com (2603:10b6:a03:333::11) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.3825.19 via Frontend
 Transport; Wed, 3 Feb 2021 02:16:47 +0000
X-MS-PublicTrafficType: Email
X-MS-Office365-Filtering-Correlation-Id: 16d09636-a85e-45d7-d614-08d8c7e9c2a9
X-MS-TrafficTypeDiagnostic: SA2PR10MB4793:
X-Microsoft-Antispam-PRVS: 
 <SA2PR10MB4793CC9BD914977039054A7581B49@SA2PR10MB4793.namprd10.prod.outlook.com>
X-MS-Oob-TLC-OOBClassifiers: OLM:4714;
X-MS-Exchange-SenderADCheck: 1
X-Microsoft-Antispam: BCL:0;
X-Microsoft-Antispam-Message-Info: 
 nfT5MHn40KNgzEEKfz978HxkPe3ohazgL+rh0G6363uvz92THc28Q+bW81ESB5w7X6gRBE57h+cQKzHtCP4YZoppCOJp7Vr71xN/1fny9pFz97BqiBo+d6u8gl4Sv5Boi7vNZ0kuanOcWXb6W9sLcg/60hrVIpBOeogunCJ64z9iONKygU78uU6qaJRhy6cd1Rr4NyMK0/MAcLt2bkyRi7Gur5raHoM0qZg3+GORDj2AdhN9Hw6p7myJ9zEMTo5HxlFNaIg8CsuXhFkaJ/5Macz36AadhNixyhAogpCzYjrNeFolCOZQzFdl7OOPuz7gJNLGNz8rbIspIuHzcsUz29RzA/R7wU/176XKEsi5l5G+jp63onZYPUimxcqFs0Pvr0iDowIH1Vz6waCqMdAmoailBgFWs/48f+od11jdq9/ouH4uMGbaf2r5sJswdBcg4G0AYT+2Ha1Isjyq2Ajo0M6Iev0UXKNzXd0Y+vmBO6HCs+H0xXKwOtk6/IWZrROn40LVvZYTeO0mHfbRVTTCqA==
X-Forefront-Antispam-Report: 
 CIP:255.255.255.255;CTRY:;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:SN6PR10MB3021.namprd10.prod.outlook.com;PTR:;CAT:NONE;SFS:(366004)(39860400002)(376002)(346002)(396003)(136003)(316002)(6666004)(36756003)(8676002)(66946007)(16526019)(66476007)(6486002)(7696005)(83380400001)(4326008)(6916009)(2906002)(86362001)(5660300002)(956004)(186003)(1076003)(44832011)(52116002)(8936002)(26005)(478600001)(2616005)(66556008);DIR:OUT;SFP:1101;
X-MS-Exchange-AntiSpam-MessageData: 
 HhnyPd8XCP72VRI8ZDPT4OgpzgjlTHd6QUcZm7yf2xGbwNo7gwKBZGangOEjVmVcmL2+5M/mqm1PBqxDT+B6HToA/OXFWToiIqWDSvzBjl0Lt4M8IgBGhMJM+4HKi+JY+e0jq8Q2J3EHdPqqE48K8wSQdCOStV3scgsQoV3iGY215lHggsJSBaofBDEMQkjiTyJdzL6bT62CXnfnohSW/DhEsE1mVjyFf7hqDyQYDUGFa+FwVtx/lj4S0dMtcrsPCRnA4SGRmfJIEgjRDpJz94sDSdlUjhVY8TCe+9d9ZbtixBxsV4H8NfPsxrWL0ev7fhheQw1t0H1qceQG9vjJlK4ipumYZKHxKfaPKjNfehpoDtAP64Orb0mdGb4SyGeUPLjsyz3Z6gTEtZaN4mFJfAdCtCEO8btlEpSg2PQk2ASoSuhKbwbc99gJrWRmYQiQKuJW00rCRDF/5qaqNY1W18xCvA4ioxbyoDG0w245SS8e4qBl0enWAyiPo46a7ZQKtff1jgplE6WowoREW2ao63iPfT/5jBDmaE4QFagf7X0liHGiDo3UWHghw7bwFK/p970UzrZR8hXGJ5MY/YU3Hgjv5YCLoOG6KAV0NLVWEGHci+AEPy2E0JmnGwcplxba6q40tCYWaznJdoWRUgC3xZRiuXOhME+yxsHWr+K0r2Dz4kZbI6qTsRBLAAh/Tdq+oWp6nI45o+JTP5PxuU2+naelCSju3lm5cyvRq1RLfniFOGivdmYFCfQlCAGWKY3fjSpb06eKMqXXqjoXf3jq+OCfo/aDpr1PtXTGr/9I/svuBDMNSrtPKzv16OYhm8R/ebzWGI27XGP5Ku46p6Lt0OHtAbaspAa+jxkOi8L6nUfsSu8rrQe/j32oM2ORgVsvxCelUISqqd5uzmgkiKYD2hP41RXr+FCHSrXnCpWooHW4M0elyL7UaZLLsfuKUInvjeGpen8cpc+ZRcgncuLzz/2GD/eCS+ZLomam4sN/AbwkOFIrsN0vBYaSHSaSwlYcLA/JMpd1srzGbHPjJmmBRBZT+nhl4ch71oCUIuCDMqTTsdhdC/iDmYQs9hi24O5g
X-OriginatorOrg: oracle.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 16d09636-a85e-45d7-d614-08d8c7e9c2a9
X-MS-Exchange-CrossTenant-AuthSource: SN6PR10MB3021.namprd10.prod.outlook.com
X-MS-Exchange-CrossTenant-AuthAs: Internal
X-MS-Exchange-CrossTenant-OriginalArrivalTime: 03 Feb 2021 02:16:48.2421
 (UTC)
X-MS-Exchange-CrossTenant-FromEntityHeader: Hosted
X-MS-Exchange-CrossTenant-Id: 4e2c6054-71cb-48f1-bd6c-3a9705aca71b
X-MS-Exchange-CrossTenant-MailboxType: HOSTED
X-MS-Exchange-CrossTenant-UserPrincipalName: 
 m9EeoeQEvyu+PxRfIP5ym7yWJO/EljHlBVCVY3HAsNMVd6x2hqFYzy8PcVuJtDlQy+/1bMgQIsfL23DTp7a5VXYw4nCFQBXhdr7NFXYPIV4=
X-MS-Exchange-Transport-CrossTenantHeadersStamped: SA2PR10MB4793
X-Proofpoint-Virus-Version: vendor=nai engine=6200 definitions=9883
 signatures=668683
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 mlxscore=0
 mlxlogscore=999 phishscore=0
 spamscore=0 suspectscore=0 malwarescore=0 adultscore=0 bulkscore=0
 classifier=spam adjust=0 reason=mlx scancount=1 engine=8.12.0-2009150000
 definitions=main-2102030007
X-Proofpoint-Virus-Version: vendor=nai engine=6200 definitions=9883
 signatures=668683
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 bulkscore=0
 adultscore=0
 priorityscore=1501 impostorscore=0 malwarescore=0 clxscore=1015
 spamscore=0 lowpriorityscore=0 phishscore=0 mlxlogscore=999 mlxscore=0
 suspectscore=0 classifier=spam adjust=0 reason=mlx scancount=1
 engine=8.12.0-2009150000 definitions=main-2102030007
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

Add helper functions to set host RFLAGS.TF immediately before the VMRUN
instruction. These will be used  by the next patch to test Single Stepping
on the VMRUN instruction from the host's perspective.

Signed-off-by: Krish Sadhukhan <krish.sadhukhan@oracle.com>
---
 x86/svm.c       | 24 ++++++++++++++++++++++--
 x86/svm.h       |  3 +++
 x86/svm_tests.c |  1 -
 3 files changed, 25 insertions(+), 3 deletions(-)

diff --git a/x86/svm.c b/x86/svm.c
index a1808c7..547f62a 100644
--- a/x86/svm.c
+++ b/x86/svm.c
@@ -179,6 +179,17 @@ void vmcb_ident(struct vmcb *vmcb)
 	}
 }
 
+static bool ss_bp_on_vmrun = false;
+void set_ss_bp_on_vmrun(void)
+{
+	ss_bp_on_vmrun = true;
+}
+
+void unset_ss_bp_on_vmrun(void)
+{
+	ss_bp_on_vmrun = false;
+}
+
 struct regs regs;
 
 struct regs get_regs(void)
@@ -215,6 +226,12 @@ struct svm_test *v2_test;
                 "mov regs, %%r15\n\t"           \
                 "mov %%r15, 0x1f8(%%rax)\n\t"   \
                 LOAD_GPR_C                      \
+                "cmpb $0, %[ss_bp]\n\t"         \
+                "je 1f\n\t"                     \
+                "pushf; pop %%r8\n\t"           \
+                "or $0x100, %%r8\n\t"           \
+                "push %%r8; popf\n\t"           \
+                "1: "                           \
                 "vmrun %%rax\n\t"               \
                 SAVE_GPR_C                      \
                 "mov 0x170(%%rax), %%r15\n\t"   \
@@ -234,7 +251,8 @@ int svm_vmrun(void)
 	asm volatile (
 		ASM_VMRUN_CMD
 		:
-		: "a" (virt_to_phys(vmcb))
+		: "a" (virt_to_phys(vmcb)),
+		[ss_bp]"m"(ss_bp_on_vmrun)
 		: "memory", "r15");
 
 	return (vmcb->control.exit_code);
@@ -253,6 +271,7 @@ static void test_run(struct svm_test *test)
 	do {
 		struct svm_test *the_test = test;
 		u64 the_vmcb = vmcb_phys;
+
 		asm volatile (
 			"clgi;\n\t" // semi-colon needed for LLVM compatibility
 			"sti \n\t"
@@ -266,7 +285,8 @@ static void test_run(struct svm_test *test)
 			"=b" (the_vmcb)             // callee save register!
 			: [test] "0" (the_test),
 			[vmcb_phys] "1"(the_vmcb),
-			[PREPARE_GIF_CLEAR] "i" (offsetof(struct svm_test, prepare_gif_clear))
+			[PREPARE_GIF_CLEAR] "i" (offsetof(struct svm_test, prepare_gif_clear)),
+			[ss_bp]"m"(ss_bp_on_vmrun)
 			: "rax", "rcx", "rdx", "rsi",
 			"r8", "r9", "r10", "r11" , "r12", "r13", "r14", "r15",
 			"memory");
diff --git a/x86/svm.h b/x86/svm.h
index a0863b8..d521972 100644
--- a/x86/svm.h
+++ b/x86/svm.h
@@ -391,6 +391,9 @@ void vmcb_ident(struct vmcb *vmcb);
 struct regs get_regs(void);
 void vmmcall(void);
 int svm_vmrun(void);
+int svm_vmrun1(void);
+void set_ss_bp_on_vmrun(void);
+void unset_ss_bp_on_vmrun(void);
 void test_set_guest(test_guest_func func);
 
 extern struct vmcb *vmcb;
diff --git a/x86/svm_tests.c b/x86/svm_tests.c
index 29a0b59..7bf3624 100644
--- a/x86/svm_tests.c
+++ b/x86/svm_tests.c
@@ -2046,7 +2046,6 @@ static void basic_guest_main(struct svm_test *test)
 {
 }
 
-
 #define SVM_TEST_REG_RESERVED_BITS(start, end, inc, str_name, reg, val,	\
 				   resv_mask)				\
 {									\

From patchwork Wed Feb  3 01:28:42 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Krish Sadhukhan <krish.sadhukhan@oracle.com>
X-Patchwork-Id: 12063193
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-18.7 required=3.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,
	INCLUDES_PATCH,MAILING_LIST_MULTI,MSGID_FROM_MTA_HEADER,SPF_HELO_NONE,
	SPF_PASS,URIBL_BLOCKED,USER_AGENT_GIT autolearn=ham autolearn_force=no
	version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 8844AC433E0
	for <kvm@archiver.kernel.org>; Wed,  3 Feb 2021 02:17:56 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 300E364F4A
	for <kvm@archiver.kernel.org>; Wed,  3 Feb 2021 02:17:56 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S232005AbhBCCRi (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Tue, 2 Feb 2021 21:17:38 -0500
Received: from userp2120.oracle.com ([156.151.31.85]:39128 "EHLO
        userp2120.oracle.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S230087AbhBCCRh (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 2 Feb 2021 21:17:37 -0500
Received: from pps.filterd (userp2120.oracle.com [127.0.0.1])
        by userp2120.oracle.com (8.16.0.42/8.16.0.42) with SMTP id
 1132EEPd076995;
        Wed, 3 Feb 2021 02:16:51 GMT
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=oracle.com;
 h=from : to : cc :
 subject : date : message-id : in-reply-to : references :
 content-transfer-encoding : content-type : mime-version;
 s=corp-2020-01-29; bh=gNcp1na4roEqmZPxwEkHPczpoNecuvyQsv2WBpPdlWw=;
 b=UP1iVKyPuQ+JBUU5qx+XQJ36VCy4CNJTZBUVRfb7TF4xxHNWx98xGwKR/rGG4Lxw1DO5
 OAiH5+ylLH0d6IxlAH1q8I0BzLUrEJcsh8pr5+SI0eA9W/ZdoedykaR6xrPke8Ybyeni
 mB+wyOIzTd+dX/R5Rx4T0e0e5hknb+lwgWW8XJOq6i70wGHGtWePga3T4anLV5sI/pNX
 j70PqEzpevno8NebL314B353rK4Fju11WuvUhh8NbHivALv4fSu/PCtj8FUQaCfaBwk2
 epKYI/DEpev7uNTkm8qsJF6DTmWspm6H0uR7VNXZBa/t25GA1Aun3abyMmKgMGGbsI3f lA==
Received: from userp3030.oracle.com (userp3030.oracle.com [156.151.31.80])
        by userp2120.oracle.com with ESMTP id 36dn4wkkat-1
        (version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Wed, 03 Feb 2021 02:16:50 +0000
Received: from pps.filterd (userp3030.oracle.com [127.0.0.1])
        by userp3030.oracle.com (8.16.0.42/8.16.0.42) with SMTP id
 1132Eand110778;
        Wed, 3 Feb 2021 02:16:50 GMT
Received: from nam10-bn7-obe.outbound.protection.outlook.com
 (mail-bn7nam10lp2102.outbound.protection.outlook.com [104.47.70.102])
        by userp3030.oracle.com with ESMTP id 36dhcxqdsa-3
        (version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Wed, 03 Feb 2021 02:16:50 +0000
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=LLDdm6iPkZQHEYdN3DwLDYCj4acspCIBMWcz28zFpvFuOEe1ZZ0rM82moN9DyR5pmF2Bi2QsqwqUuzUOON9G9KX04m/Oh7Jbj/S309SOIPQ3nxPt9N9ZafAsE2cGwupZ/UQ6Kr4Dax9tYBNgxCvEqSS/bJDYmcbqMn7SzN1RNCyOKFX1BA+Zk2DFsAd1pvzRqC+Wnml6/jEBPVZZwqwAM9wR9U0TPuEn+40LvtOYZYoxckKI8BU5p8g8CsaCrBXoM38mKbp/KORUzJOcyW9inKGU1vBVSmrs2SOU4FBCm9ZL8fn/mQeRNpigqjIkTqzWFJbyoIPTzx+erq+LCLfkMQ==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=gNcp1na4roEqmZPxwEkHPczpoNecuvyQsv2WBpPdlWw=;
 b=QO33GORZFp4+zTgnEdH3MkjwubMgep7BBEm4ha8UvFMmIaGdf2maHlVyWmJ8lssmdGkHuPdLOMzmiLcphL9sqdTr+C1Bp1hjAeo+76WgokY5rH32k0V20I+nridkUlt+9tOoWKk3uM6nLVD/RD84pBaI90IgBPTD4KMpwysymrjUYc3cK35OE4Sdp3DVEwe1XOGHWA9USKms6VmwUdXynBMdKqU5apzupRhnCfxROzxh4V4DzHWKFS1RW4UPFvMXBelPfbOrvwbwK85cMGbjV/RuGU/v9aYtK3mOsWpsmtv13Szjf+9I5DEVs5+cIFKdP19j43AuaIXhfOKYQvm81g==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=oracle.com; dmarc=pass action=none header.from=oracle.com;
 dkim=pass header.d=oracle.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=oracle.onmicrosoft.com; s=selector2-oracle-onmicrosoft-com;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=gNcp1na4roEqmZPxwEkHPczpoNecuvyQsv2WBpPdlWw=;
 b=ITwm2+BF0kWGfJBBPaaNF8dE+ZbGDGbrySPBzqKKqIgSOrYgAKsVqulR1R4oAWaWkKqEJUj7+Sh2OSYoLKf3hjuy5parKRJfwhkmqWPfNKp/pdIidmrrjFSL4e71o4j1LHowiHHXPTiBwMXa1v7c+jhTczeTPO1BMk1NUsctfiM=
Authentication-Results: vger.kernel.org; dkim=none (message not signed)
 header.d=none;vger.kernel.org; dmarc=none action=none header.from=oracle.com;
Received: from SN6PR10MB3021.namprd10.prod.outlook.com (2603:10b6:805:cc::19)
 by SA2PR10MB4793.namprd10.prod.outlook.com (2603:10b6:806:110::24) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.3805.17; Wed, 3 Feb
 2021 02:16:49 +0000
Received: from SN6PR10MB3021.namprd10.prod.outlook.com
 ([fe80::1871:3741:cc17:bcf7]) by SN6PR10MB3021.namprd10.prod.outlook.com
 ([fe80::1871:3741:cc17:bcf7%7]) with mapi id 15.20.3825.020; Wed, 3 Feb 2021
 02:16:49 +0000
From: Krish Sadhukhan <krish.sadhukhan@oracle.com>
To: kvm@vger.kernel.org
Cc: pbonzini@redhat.com, jmattson@google.com, seanjc@google.com
Subject: [PATCH 3/3] nSVM: Test effect of host RFLAGS.TF on VMRUN
Date: Tue,  2 Feb 2021 20:28:42 -0500
Message-Id: <20210203012842.101447-4-krish.sadhukhan@oracle.com>
X-Mailer: git-send-email 2.25.4
In-Reply-To: <20210203012842.101447-1-krish.sadhukhan@oracle.com>
References: <20210203012842.101447-1-krish.sadhukhan@oracle.com>
X-Originating-IP: [138.3.200.29]
X-ClientProxiedBy: SJ0PR03CA0096.namprd03.prod.outlook.com
 (2603:10b6:a03:333::11) To SN6PR10MB3021.namprd10.prod.outlook.com
 (2603:10b6:805:cc::19)
MIME-Version: 1.0
X-MS-Exchange-MessageSentRepresentingType: 1
Received: from ban25x6uut29.us.oracle.com (138.3.200.29) by
 SJ0PR03CA0096.namprd03.prod.outlook.com (2603:10b6:a03:333::11) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.3825.19 via Frontend
 Transport; Wed, 3 Feb 2021 02:16:48 +0000
X-MS-PublicTrafficType: Email
X-MS-Office365-Filtering-Correlation-Id: 88b90c1f-1723-4982-a76c-08d8c7e9c352
X-MS-TrafficTypeDiagnostic: SA2PR10MB4793:
X-Microsoft-Antispam-PRVS: 
 <SA2PR10MB4793DFCFF0EFF3672B435A9281B49@SA2PR10MB4793.namprd10.prod.outlook.com>
X-MS-Oob-TLC-OOBClassifiers: OLM:6790;
X-MS-Exchange-SenderADCheck: 1
X-Microsoft-Antispam: BCL:0;
X-Microsoft-Antispam-Message-Info: 
 nXeS5UdIfxRTE9egvs8EyMpo2StQsoS6+y1vvP1WEwV3yhGm3lvLgdlchAhcwRz85pw0hE1Tm1Ta1fNehsZb7zgMtcYXrxKkFlnz3rLX1I1CkbwHvoAsbkWNZawdTMRedsc1WNYaZY6PaDe9+rXu04LiUfWVkGawNzu+khPr/vZBQtuEXiw0STVFgfAllzy4diAb6gtpp1L39rxAydOym6M3ELwpCWZIm9KFw/ruZgnjfLGTetMVzHRachIZbhEOYNbKTvUNAq+G/DLFXnuDLuEomLSHXdBUAYhFQUi5o2ZgakWJ/8FKBwK2dLSFt1TJUFw5z1R5/l/h40/lHZ2lQpyR2YfiRL5WYVvjVQToOh4gM28W+e5hW76xZoZJsvIzR6dSDg+KfsiOG3mnCByZcUtUdatvHgqB45bPZm8PUy4NbGJcydeop4Xev090o05RXtpP09k7tlM3sXnqIzKPEMamM1lr32Q4D4SQfNqHfhalmF1vnRDjywBausDSV0oEabGS7l7VpZa/yLZASUwnYA==
X-Forefront-Antispam-Report: 
 CIP:255.255.255.255;CTRY:;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:SN6PR10MB3021.namprd10.prod.outlook.com;PTR:;CAT:NONE;SFS:(366004)(39860400002)(376002)(346002)(396003)(136003)(316002)(6666004)(36756003)(8676002)(66946007)(16526019)(66476007)(6486002)(7696005)(83380400001)(4326008)(6916009)(2906002)(86362001)(5660300002)(956004)(186003)(1076003)(44832011)(52116002)(8936002)(26005)(478600001)(2616005)(66556008);DIR:OUT;SFP:1101;
X-MS-Exchange-AntiSpam-MessageData: 
 pNFvZADbnM6GG1a/ZGIkVarUgGhDS1ivfafn7vLY3xHeQZkdBpEoUpKetFEaHqIe0QNJGCWtM+KwAE12mziwfnlSFBD1ktFDM+v7VkHU3/RQP5r2j3h6iZyvYdHkFkcXal3Ko2RJo1Ch83hOFdf035MpilDF+yG8iN0xaV21ihN3GaL1q2pHF3cwRCHyUd+FxzEGGIbj8Pqefk+Hj29XEkucmMJruPvj41Q8JdBzFr2eHCBxUyXcg6eVUhGKMk00SIbocrfWzK+8Cth/9/F0XbvvZYvUvevM8vbtTsBLv8ebkIkJfY7RXpwmLSNn2ThlJXvKDVwP5D4euFUiKBiXmqA18nAjtiC6xeUNckxD8nJRS7YH99ZybIarn5ngWXOJHytJf56z/XRQhndobgs3EgJLi3mMkIxGOY7APC9cg41fopFYHiQ9d0pj2L2iLxPUp7PugBQ0BfLnQ/ibJohVh9GYWCpb2BMaLbs7iSD+LRi/9axOVvUZhz/Cb51bP0CBS52w0xD+33Uw/rnZrXhZYZEFuLeaE940cDUJ1DTAfj1/J01mljMdDIk+2d1s5hE9LM/gvzeECyB/ZSWSRtnE9tkB57nGT6/ZJRnN9nE6yJVi7nIR+UU8DZ67EXas0zFMu3hh/J3jUU9QYi558eogQn65MkMZUZn6zrsdCZakUuDsX9tVl7JL9jok+fLcgsuHp6kbDOWYYIbgbdua4O9mTvYQbNEqxpBZChI9s07vOmV9gXJsvOazIqyer/pqVchN7emIEOpgJ08CqxTYiwUoY2sYEAmamR5YzBLysSZi7jvYpyj+lsn9LpWLAanXSdeszFlJKRe1W6gOKbiAzkf5rckJSaS1Wp2BREl03Oqg9VUQMxDZkEQGlQsrEmF3xmu8DXsdMWMB0EliF9UkLkY0sWYBd8vJ7S6UVqdE+h519eSY+LtEgFYUUGygN4BrFEosdMudW3RZW0MuBNkV6lm/zgTG+OJkD+H2Do/3fphKmHT6wIL6DsLDS49NBg9qa+zFzerGQ3K/dryyLWjhDfDJHjX2UiWadlIdOxl16CJyYDGYf78Sb4oBXltEzO3Q6GkJ
X-OriginatorOrg: oracle.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 88b90c1f-1723-4982-a76c-08d8c7e9c352
X-MS-Exchange-CrossTenant-AuthSource: SN6PR10MB3021.namprd10.prod.outlook.com
X-MS-Exchange-CrossTenant-AuthAs: Internal
X-MS-Exchange-CrossTenant-OriginalArrivalTime: 03 Feb 2021 02:16:49.3745
 (UTC)
X-MS-Exchange-CrossTenant-FromEntityHeader: Hosted
X-MS-Exchange-CrossTenant-Id: 4e2c6054-71cb-48f1-bd6c-3a9705aca71b
X-MS-Exchange-CrossTenant-MailboxType: HOSTED
X-MS-Exchange-CrossTenant-UserPrincipalName: 
 +ff5FgoAqil06PntecgQIHfcvF7GRl/nc2Z8N98eGdPrrdVuXyMH+cEmS4OIFBkiD1llixtrCMA/h7Z5G/dPwWR1b82AOet0ovNpjcMd4wc=
X-MS-Exchange-Transport-CrossTenantHeadersStamped: SA2PR10MB4793
X-Proofpoint-Virus-Version: vendor=nai engine=6200 definitions=9883
 signatures=668683
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 mlxscore=0
 mlxlogscore=999 phishscore=0
 spamscore=0 suspectscore=0 malwarescore=0 adultscore=0 bulkscore=0
 classifier=spam adjust=0 reason=mlx scancount=1 engine=8.12.0-2009150000
 definitions=main-2102030007
X-Proofpoint-Virus-Version: vendor=nai engine=6200 definitions=9883
 signatures=668683
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 impostorscore=0
 lowpriorityscore=0
 spamscore=0 priorityscore=1501 suspectscore=0 phishscore=0 mlxlogscore=999
 malwarescore=0 clxscore=1015 bulkscore=0 adultscore=0 mlxscore=0
 classifier=spam adjust=0 reason=mlx scancount=1 engine=8.12.0-2009150000
 definitions=main-2102030007
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

According to section "VMRUN and TF/RF Bits in EFLAGS" in AMD APM vol 2,

	"From the host point of view, VMRUN acts like a single instruction,
	 even though an arbitrary number of guest instructions may execute
	 before a #VMEXIT effectively completes the VMRUN. As a single
	 host instruction, VMRUN interacts with EFLAGS.TF like ordinary
	 instructions. EFLAGS.TF causes a #DB trap after the VMRUN completes
	 on the host side (i.e., after the #VMEXIT from the guest).

Signed-off-by: Krish Sadhukhan <krish.sadhukhan@oracle.com>
---
 x86/svm_tests.c | 93 +++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 93 insertions(+)

diff --git a/x86/svm_tests.c b/x86/svm_tests.c
index 7bf3624..73bffe6 100644
--- a/x86/svm_tests.c
+++ b/x86/svm_tests.c
@@ -2002,6 +2002,96 @@ static bool init_intercept_check(struct svm_test *test)
     return init_intercept;
 }
 
+/*
+ * Setting host EFLAGS.TF causes a #DB trap after the VMRUN completes on the
+ * host side (i.e., after the #VMEXIT from the guest).
+ *
+ * [AMD APM]
+ */
+static volatile u8 host_rflags_guest_flag = 0;
+static volatile u8 host_rflags_isr_flag = 0;
+static void ss_bp_isr(struct ex_regs *r)
+{
+	host_rflags_isr_flag = (host_rflags_guest_flag == 1) ? 2 : 1;
+	r->rflags &= ~X86_EFLAGS_TF;
+}
+
+static void host_rflags_prepare(struct svm_test *test)
+{
+	default_prepare(test);
+	handle_exception(DB_VECTOR, ss_bp_isr);
+	host_rflags_guest_flag = host_rflags_isr_flag = 0;
+	set_test_stage(test, 0);
+}
+
+static void host_rflags_test(struct svm_test *test)
+{
+	while (1) {
+		if (get_test_stage(test) > 0)
+			host_rflags_guest_flag =
+			    (host_rflags_isr_flag == 1) ? 2 : 1;
+		vmmcall();
+		if (get_test_stage(test) == 3)
+			break;
+	}
+}
+
+static bool host_rflags_finished(struct svm_test *test)
+{
+	switch (get_test_stage(test)) {
+	case 0:
+		if (vmcb->control.exit_code != SVM_EXIT_VMMCALL &&
+		    host_rflags_isr_flag != 0 && host_rflags_guest_flag != 0) {
+			report(false, "Unexpected VMEXIT. Exit reason 0x%x",
+			    vmcb->control.exit_code);
+			return true;
+		}
+		/*
+		 * Setting host EFLAGS.TF not immediately before VMRUN, causes
+		 * #DB trap before first guest instruction is executed
+		 */
+		write_rflags(read_rflags() | X86_EFLAGS_TF);
+		vmcb->save.rip += 3;
+		break;
+	case 1:
+		if (vmcb->control.exit_code != SVM_EXIT_VMMCALL &&
+		    host_rflags_isr_flag != 1 && host_rflags_guest_flag != 2) {
+			report(false, "Unexpected VMEXIT. Exit reason 0x%x",
+			    vmcb->control.exit_code);
+			return true;
+		}
+		host_rflags_guest_flag = host_rflags_isr_flag = 0;
+		vmcb->save.rip += 3;
+		/*
+		 * Setting host EFLAGS.TF immediately before VMRUN, causes #DB
+		 * trap after VMRUN completes on the host side (i.e., after
+		 * VMEXIT from guest).
+		 */
+		set_ss_bp_on_vmrun();
+		break;
+	case 2:
+		if (vmcb->control.exit_code != SVM_EXIT_VMMCALL &&
+		    host_rflags_isr_flag != 2 && host_rflags_guest_flag != 1) {
+			report(false, "Unexpected VMEXIT. Exit reason 0x%x",
+			    vmcb->control.exit_code);
+			return true;
+		}
+		host_rflags_guest_flag = host_rflags_isr_flag = 0;
+		vmcb->save.rip += 3;
+		unset_ss_bp_on_vmrun();
+		break;
+	default:
+		return true;
+	}
+	inc_test_stage(test);
+	return get_test_stage(test) == 4;
+}
+
+static bool host_rflags_check(struct svm_test *test)
+{
+	return get_test_stage(test) == 3;
+}
+
 #define TEST(name) { #name, .v2 = name }
 
 /*
@@ -2491,6 +2581,9 @@ struct svm_test svm_tests[] = {
     { "svm_init_intercept_test", smp_supported, init_intercept_prepare,
       default_prepare_gif_clear, init_intercept_test,
       init_intercept_finished, init_intercept_check, .on_vcpu = 2 },
+    { "host_rflags", default_supported, host_rflags_prepare,
+      default_prepare_gif_clear, host_rflags_test,
+      host_rflags_finished, host_rflags_check },
     TEST(svm_cr4_osxsave_test),
     TEST(svm_guest_state_test),
     TEST(svm_vmrun_errata_test),
